### PR TITLE
Create weekly event recurring instance

### LINKED ISSUE #1583: Fix Recurring Events
---
### Note

This is a very important feature to implement. 
1. We need someone with either demonstrable:
    1. experience with GraphQL
    2. activity in their GitHub account with a focus on TypeScript
1. This is not an issue for a first time attempt at the technology. 
    1. The first PR needs to be created within a week
    1. All PRs must be completed by the end of February
1. Read through the comments in the closed PR about the progress and why the approach was selected. There are further details below.

In summary:
1. This is not for a newbie
1. This is not for someone with limited time over the next month
---

**Describe the bug**
The creation of a recurring event only creates a single event in the API clients.

**To Reproduce**

1. Create a recurring event
3. Only a single instance is visible

**Expected behavior**

1. You'll need to update or create new mutations that include start and end date for the event recurrence, not just the event itself
1. There will need to be a query created or updated that would return all the calendar entries for a particular time frame including the recurring ones.
1. We will need the ability to edit any recurrence of the event at any time that the event occurs. The edits would need to be valid for:
    1. All instances in the series
    2. Just the instance in question
    4. The instance in question and all those in future
1. Administrators will need to be able to track attendance for any type of event. Therefore you will need to think of ways to create permanence for historical recurring events for this purpose
1. We don't want the creation of an infinite number of event calendar entries if a recurring event has no end date
1. Other functionality must not be affected
1. You must write tests to validate your work

**Actual behavior**
- See above

**Screenshots**
- N/A

**Additional details**
1. Work with @xoldyckk on evaluating your solution here and in the subsequent PR
1. This is a prerequisite for our menu of project ideas for GSoC 2024
1.  When this issue is resolved we'll need to create another one to support custom recurrences, such as `3rd week of the month`, similar to what is available in Google Calendar
1. We almost completed the work for Weekly recurring when the assignee @Community-Programmer was unable to continue due to exam time constraints. 
    1. The first PR for this issue will need to build on the work in this abandoned PR as it was almost approved for merging. 
        1. https://github.com/PalisadoesFoundation/talawa-api/pull/1738
    1. You will need to submit a plan as described here
         1. https://github.com/PalisadoesFoundation/talawa-api/pull/1738#issuecomment-1910724294
1. You will need to split this issue into multiple PRs, each with custom options provided by the RRULE package
    1. Weekly recurring
    2. Daily recurring
    3. Monthly recurring
    5. Annually recurring

**Potential internship candidates**
- Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by Community-Programmer**
**What kind of change does this PR introduce?**
Fixed the bug on creating recurring event only single event is created

**Issue Number:**

Fixes #1583 

**Did you add tests for your changes?**
yes

**Snapshots/Videos:**

Working after making changes

https://github.com/PalisadoesFoundation/talawa-api/assets/76515568/8e9b133c-42ad-4b62-a352-7ce03914d8aa

https://github.com/PalisadoesFoundation/talawa-api/assets/76515568/11ec4c99-7a2c-4e7e-91f3-bbb231e0cc68

**If relevant, did you update the documentation?**

No

**Summary**

- [ ] Task-1 : Create a recurring instance defaulting to a week

**Implementation Details:**

**Validation:**

No other functionalities is affected by changes made in the `createEvent ` mutation
All test files passed

image
image

**Does this PR introduce a breaking change?**

Yes

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

**Comment by palisadoes** [comment]
1. Does this also cover the case for events that are not all day?
2. That will be the more commonly used scenario and should also be a part of this PR

**Comment by Community-Programmer** [comment]
image

Yes, it does cover that, but it needs to be configured accordingly on the frontend. The `startTime` and `endTime` need to be added for creating an event that is not all day and `allDay` set to false. I think it's mostly related to the frontend. If any changes are required on backend, they can be done in the next pull request for editing/managing events.

**Comment by palisadoes** [comment]
Make the front end modifications so that this non-all day can be done as the next PR in Talawa Admin. We'll need to populate the sample DB with more realistic data and the add function alone will be enough for that, and like you said, work on editing can come later.

**Comment by palisadoes** [comment]
Please add tests for this. We need to be sure that it's bug free at every stage as there will be dozens of stages

**Comment by Community-Programmer** [comment]
> Make the front end modifications so that this non-all day can be done as the next PR in Talawa Admin. We'll need to populate the sample DB with more realistic data and the add function alone will be enough for that, and like you said, work on editing can come later.

To ensure clarity, could you confirm if the front-end modifications for non-all-day events should be implemented directly within the Talawa Admin interface? If so, I'll proceed with the necessary changes accordingly

**Comment by palisadoes** [comment]
When someone creates a non all day event and makes it recurring just add text to say that it will be weekly. No choice.

When we have other recurring frequency options then we can include @rishav-jha-mech and @aashimawadhwa for design guidance.

**Comment by Community-Programmer** [comment]
@palisadoes 

Alternatively, we can choose to hold the PR for Talawa Admin until we implement options for users to choose the recurrence frequency. We can merge the current PR as is, and plan the next PR as `Creating Monthly, Weekly, Yearly Recurrence`, followed by `Editing/Modifying Events`. This will allow me some time to work on them consecutively.

Let's complete the backend functionality first then we can design the frontend according for all recurring.all-day.non all-day events

Please let me know your thoughts on this approach and if it aligns with the project timeline.

**Comment by palisadoes** [comment]
1. OK but write tests for this PR. We don't want to replace one broken feature with another.
1. Make a separate PR for each frequency tied to the original issue. Each with tests.
1. Remember daily / weekdays, and tests
1. When each API PR is complete, we'll create an issue for the front end for others to implement it. That way we'll really know whether it works.

**Comment by palisadoes** [comment]
1. Will this PR support non all-day events?
1. Will it support every x weeks?

It must for completeness

**Comment by Community-Programmer** [comment]
OK for sure give me some days i will complete and all the functionaillities with proper testing , will ensure that it totally replicate google calendar

> 1. OK but write tests for this PR. We don't want to replace one broken feature with another.
> 2. Make a separate PR for each frequency tied to the original issue. Each with tests.
> 3. Remember daily / weekdays, and tests
> 4. When each API PR is complete, we'll create an issue for the front end for others to implement it. That way we'll really know whether it works.

Okay, for sure. Give me some days/weeks; I will complete all the functionalities with proper testing. I will ensure that it totally replicates Google Calendar

**Comment by Community-Programmer** [comment]
> 1. Will this PR support non all-day events?
> 2. Will it support every x weeks?
> 
> It must for completeness
Yes, as mentioned, for non-all day events, start time and end time need to be passed with allDay set to false. Additionally, for every x weeks, as already demonstrated in the video, it will create n number of events from the starting date to the ending date, with the day  at the start date

**Comment by Community-Programmer** [comment]
@palisadoes @xoldyckk 

we need feature something like this right?

image


- Once
- Daily
- weekly on x day
- Monthly on x day
- Annually on x day

is any other feature that need to be added?

**Comment by palisadoes** [comment]
If you look at custom, they also have day X of the Nth week of the month based on the current calendar. I guess that's similar to the `Last Sunday` option in the drop down, so add that too.

Custom also allows you to select multiple days in the week, like weekly on Tuesday and Thursday. That and other customizations can come later.

Please proceed with the suggested addition above.

**Comment by Community-Programmer** [comment]
> If you look at custom, they also have day X of the Nth week of the month based on the current calendar. I guess that's similar to the `Last Sunday` option in the drop down, so add that too.
> 
> Custom also allows you to select multiple days in the week, like weekly on Tuesday and Thursday. That and other customizations can come later.
> 
> Please proceed with the suggested addition above.

Okay, got it. Give me some time so that I can complete all of the above-mentioned features with proper test cases

Thank You

**Comment by palisadoes** [comment]
Be careful, some months like this one, have five weeks. This is the fifth Sunday of December. That's probably why the drop down says "Last Sunday"

This is a required check

**Comment by Community-Programmer** [comment]
> Be careful, some months like this one, have five weeks. This is the fifth Sunday of December. That's probably why the drop down says "Last Sunday"
> 
> This is a required check

Ohh yaa, thank you for pointing that out. I'll keep that in mind. Wish you a very happy New Year!

**Comment by Community-Programmer** [comment]
@palisadoes 

Please hold off on merging this PR as I need to make essential adjustments for alignment with other recurring event features (monthly, yearly, etc.). Will expedite and keep you updated.

Thank you

**Comment by palisadoes** [comment]
The reviewers will need to review this and recommend the merge

**Comment by Community-Programmer** [comment]
@palisadoes  @xoldyckk 

I have completed the development of all the features, including the creation of proper test files. I am now prepared to create a pull request (PR).

Currently, I have not initiated the PR to avoid potential merge conflicts. This decision is based on the need to create multiple branches for each feature and the possibility of having to cherry-pick commits, which could lead to conflicts during the merging process.
 
 So can you review this and suggest changes if any so that it can merged.
 
 Thank You

**Comment by palisadoes** [comment]
1. Go ahead @Community-Programmer 
2. @xoldyckk, @sumitra19jha  We'll need to give these PRs priority

**Comment by Community-Programmer** [comment]
@palisadoes 

So can I start creating PR for each features or wait for this to be reviewed and merged?
What about merging conflicts?

**Comment by palisadoes** [comment]
> @palisadoes
> 
> So can I start creating PR for each features or wait for this to be reviewed and merged? What about merging conflicts?

1. One PR per feature.
2. The merge conflicts should be greatly reduced after the first PR as people will be basing their work against your first merged PR.
3. This is also a brand new feature, which will reduce the risk further

Do you want this one reviewed now?

**Comment by Community-Programmer** [comment]
@palisadoes 

Yaa, that sounds like a great plan! Once this PR is reviewed and merged, it'll provide a solid foundation for the next steps.

In the meantime, I will create and ready  the documentation for changes I made. Having it ready will indeed be valuable for contributors looking to integrate the new feature with the Talawa-admin.

Thank You

**Comment by palisadoes** [comment]
@sumitra19jha @xoldyckk @DMills27 @ttaylor92 Please review

**Comment by DMills27** [comment]
Is it possible to have a recurring meeting without an end date in a similar manner to how Google Meets does it?

**Comment by Community-Programmer** [comment]
> Is it possible to have a recurring meeting without an end date in a similar manner to how Google Meets does it?



> Is it possible to have a recurring meeting without an end date in a similar manner to how Google Meets does it?


Yes, we do have that feature available. By default, we set the end date to the year 2100, which aligns with the date format Google Calendar uses for recurring events. This enhancement will be included in the upcoming pull request (PR) for daily, monthly, and yearly recurring events.

**Review by ttaylor92 — APPROVED** [review]

**Comment by Community-Programmer** [comment]
Below is the documentation detailing the changes I have made and plan to make. This information will be helpful for understanding the code and making it easier to integrate this feature with Talawa-admin

[Documentation](https://documentation1538.netlify.app/l)

File: src/resolvers/Mutation/createEvent.ts
Comment (sumitra19jha) [inline src/resolvers/Mutation/createEvent.ts:139]: Let say your db call for event creation is success but this call fails. Then as per your current logic the partial implementation will be accepted. Please start a transaction and make sure everything is success or everything fails.

File: src/resolvers/Mutation/createEvent.ts
Comment (sumitra19jha) [inline src/resolvers/Mutation/createEvent.ts:144]: consider using a switch statement or a strategy pattern for better extensibility. If-else will work but code quality can be improved

File: src/resolvers/Mutation/createEvent.ts
Comment (sumitra19jha) [inline src/resolvers/Mutation/createEvent.ts:204]: Why are you using a loop here rather than writing query to directly handle such case. The choice is load on database or server? I would say database because loop here in server should always be avoided until there is no other way.

**Review by palisadoes — APPROVED** [review]

File: src/resolvers/Mutation/createEvent.ts
Comment (sumitra19jha) [inline src/resolvers/Mutation/createEvent.ts:173]: This is not sufficient. createdEvent can be empty array right if startDate > endDate. Can you please tell is this condition handled in some directives and if not then how can you throw error if you have allowed client to provide this data

**Review by sumitra19jha — COMMENTED** [review]
Please update

**Comment by palisadoes** [comment]
In future also add the documentation in the code itself so that TypeDoc can parse it and put it automatically in our docs.talawa.io site

**Comment by palisadoes** [comment]
I'll also create an issue to get Talawa Admin and Mobile to start using Weekly recurring events.

**Comment by palisadoes** [comment]
@Community-Programmer Please resubmit with the recommendations from @sumitra19jha

**Comment by Community-Programmer** [comment]
@sumitra19jha 

Thank you for the review.

I have carefully considered the changes you suggested and am committed to implementing them to enhance the overall quality of the project.

File: src/resolvers/Mutation/createEvent.ts
Comment (Community-Programmer) [reply src/resolvers/Mutation/createEvent.ts:173]: we are already checking this,  if startDate > endDate we are returning user with the error that start date cannot be greater than end date

**Pull request merged**
